apps/emqx/src/emqx_router_syncer.erl

```diff
@@ -0,0 +1,328 @@
+        size => maps:size(Batch),
+        stashed => maps:size(Stash3),
+        n_add => maps:size(maps:filter(fun(_, ?ROUTEOP(A)) -> A == add end, Batch)),
+        n_delete => maps:size(maps:filter(fun(_, ?ROUTEOP(A)) -> A == delete end, Batch)),
+        prio_highest => maps:fold(fun(_, ?ROUTEOP(_, P), M) -> min(P, M) end, none, Batch),
+        prio_lowest => maps:fold(fun(_, ?ROUTEOP(_, P), M) -> max(P, M) end, 0, Batch)
+    }),
+    %% TODO: retry if error?
+    Errors = run_batch(Batch),
+    ok = send_replies(Errors, Batch),
+    NState = State#{stash := Stash3},
+    %% TODO: postpone if only ?PRIO_BG operations left?
+    case stash_empty(Stash3) of
+        true ->
+            NState;
+        false ->
+            run_batch_loop([], NState)
+    end.
+
+%%
+
+mk_batch(Stash) when map_size(Stash) =< ?MAX_BATCH_SIZE ->
+    %% This is perfect situation, we just use stash as batch w/o extra reallocations.
+    {Stash, stash_new()};
+mk_batch(Stash) ->
+    %% Take a subset of stashed operations to form a batch.
+    %% Note that stash is an unordered map, it's not a queue. The order of operations is
+    %% not preserved strictly, only loosely, because of how we start from high priority
+    %% operations and go down to low priority ones. This might cause some operations to
+    %% stay in stash for unfairly long time, when there are many high priority operations.
+    %% However, it's unclear how likely this is to happen in practice.
+    mk_batch(Stash, ?MAX_BATCH_SIZE).
+
+mk_batch(Stash, BatchSize) ->
+    mk_batch(?PRIO_HI, #{}, BatchSize, Stash).
+
+mk_batch(Prio, Batch, SizeLeft, Stash) ->
+    mk_batch(Prio, Batch, SizeLeft, Stash, maps:iterator(Stash)).
+
+mk_batch(Prio, Batch, SizeLeft, Stash, It) when SizeLeft > 0 ->
+    %% Iterating over stash, only taking operations with priority equal to `Prio`.
+    case maps:next(It) of
+        {Route, Op = ?ROUTEOP(_Action, Prio), NIt} ->
+            NBatch = Batch#{Route => Op},
+            NStash = maps:remove(Route, Stash),
+            mk_batch(Prio, NBatch, SizeLeft - 1, NStash, NIt);
+        {_Route, _Op, NIt} ->
+            %% This is lower priority operation, skip it.
+            mk_batch(Prio, Batch, SizeLeft, Stash, NIt);
+        none ->
+            %% No more operations with priority `Prio`, go to the next priority level.
+            true = Prio < ?PRIO_BG,
+            mk_batch(Prio + 1, Batch, SizeLeft, Stash)
+    end;
+mk_batch(_Prio, Batch, _, Stash, _It) ->
+    {Batch, Stash}.
+
+send_replies(Errors, Batch) ->
+    maps:foreach(
+        fun(Route, {_Action, _Prio, Ctx}) ->
+            case Ctx of
+                [] ->
+                    ok;
+                _ ->
+                    replyctx_send(maps:get(Route, Errors, ok), Ctx)
+            end
+        end,
+        Batch
+    ).
+
+replyctx_send(_Result, []) ->
+    noreply;
+replyctx_send(Result, {MRef, Pid}) ->
+    Pid ! {MRef, Result}.
```

Comment:
Nit: prefer ignoring the result in the earliest place, but not delegate to callers.
```suggestion
replyctx_send(_Result, []) ->
    ok;
replyctx_send(Result, {MRef, Pid}) ->
    _ = erlang:send(Pid, {MRef, Result}),
    ok.
```